src/System/Net/Benchmarks/Apps/Socket/Client/Program.cs

```diff
@@ -0,0 +1,247 @@
+        LogMetricPercentiles(MetricName.ConnectionEstablishment, "Connection establishment duration (ms)", metrics.SelectMany(x => x).ToList());
+    }
+
+    private async Task RunReadWriteScenario(SocketClientOptions options, CancellationToken cancellationToken)
+    {
+        var connections = new SocketClientConnection[options.Connections];
+        var tasks = new List<Task<ReadWriteMetrics>>(options.Connections);
+        for (var i = 0; i < options.Connections; i++)
+        {
+            connections[i] = await EstablishConnectionAsync(options).ConfigureAwait(false);
+            tasks.Add(ReadWriteScenario(connections[i], options));
+        }
+
+        var metrics = await Task.WhenAll(tasks).WaitAsync(cancellationToken).ConfigureAwait(false);
+        LogMetric(MetricName.Read + MetricName.Mean, "Read B/s - mean", metrics.Sum(x => x.BytesReadPerSecond));
+        LogMetric(MetricName.Write + MetricName.Mean, "Write B/s - mean", metrics.Sum(x => x.BytesWrittenPerSecond));
+
+        foreach (var connection in connections)
+        {
+            await connection.DisposeAsync();
+        }
+    }
+
+    private async Task RunRpsScenario(SocketClientOptions options, CancellationToken cancellationToken)
+    {
+        var connections = new SocketClientConnection[options.Connections];
+        var tasks = new List<Task<(double Rps, long Errors)>>(options.Connections);
+        for (var i = 0; i < options.Connections; i++)
+        {
+            connections[i] = await EstablishConnectionAsync(options).ConfigureAwait(false);
+            tasks.Add(RpsScenario(connections[i], options));
+        }
+
+        var metrics = await Task.WhenAll(tasks).WaitAsync(cancellationToken).ConfigureAwait(false);
+        LogMetric(MetricName.Rps + MetricName.Mean, "RPS - mean", metrics.Sum(x => x.Rps));
+        if (metrics.Any(x => x.Errors > 0))
+        {
+            LogMetric(MetricName.Errors, "Errors", metrics.Sum(x => x.Errors), "n0");
+        }
+
+        foreach (var connection in connections)
+        {
+            await connection.DisposeAsync();
+        }
+    }
+
+    private static async Task<List<double>> ConnectionEstablishmentScenario(Func<SocketClientOptions, Task<SocketClientConnection>> establishConnectionAsync, SocketClientOptions options)
+    {
+        var values = new List<double>((int)options.Duration.TotalMilliseconds);
+        var isWarmup = true;
+        var sw = Stopwatch.StartNew();
+        while (IsRunning)
+        {
+            if (isWarmup && !IsWarmup)
+            {
+                isWarmup = false;
+                values.Clear();
+                OnWarmupCompleted();
+            }
+            sw.Restart();
+            await using var connection = await establishConnectionAsync(options).ConfigureAwait(false);
+            await using var stream = await connection.EstablishStreamAsync(options).ConfigureAwait(false);
```

Comment:
Wait, I am a bit confused... we discussed that we need socket benchmarks that call actual Socket APIs :) so that we can have/add scenarios that are impossible to do using Stream APIs (e.g. UDP socket benchmarks)

When we've discussed NetworkStream previously, in my mind that was about an _additional_ benchmark, more intended to compare with SslStream to see the TLS overhead

Reply:
I see, looks like I got confused from our last discussion as well, and didn't actually realize I should be using direct socket APIs here, I'll change the design to use Socket APIs directly.